ml-agents/mlagents/trainers/agent_processor.py

```diff
@@ -152,12 +150,14 @@ def add_experiences(
                         )
                         del self.episode_steps[agent_id]
                         del self.episode_rewards[agent_id]
-                elif not next_info.local_done[next_idx]:
+                elif not curr_info.local_done[next_idx]:
                     self.episode_steps[agent_id] += 1
 
+            self.last_brain_info[agent_id] = curr_info
+
         if "action" in take_action_outputs:
             self.policy.save_previous_action(
-                curr_info.agents, take_action_outputs["action"]
+                previous_action.agents, take_action_outputs["action"]
```

Comment:
Why are the agent_processor and the policy not storing the previous action at the same time?

Reply:
I think the meaning of the previous action is different. In the Policy. it's the previous taken action, while in the agent_processor, last_take_action_outputs corresponds to the last seen action (in effect the current_action). 

So when we assemble an AgentExperience, there should be _two_ actions - the prev_action from the policy and the current action from the last_take_action_outputs. 

Would it be less confusing to rename `last_take_action_outputs` to `last_seen_take_action_outputs`?

Reply:
Not sure renaming would be enough to clear confusion here. I would say add a comment.

Reply:
Added a comment and a note about future directions :P